inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -0,0 +1,209 @@
+    } catch(const IE::Exception &iie) {
+        IE_THROW() << "Failed to load network to device named " << selectedDevice.deviceName
+                   << " with exception " << iie.what();
+    }
+
+    bool enablePerfCounters = false;
+    try {
+        enablePerfCounters =
+            executableNetwork.GetConfig(IE::PluginConfigParams::KEY_PERF_COUNT).as<std::string>() ==
+                IE::PluginConfigParams::YES;
+    } catch (...) {
+    }
+
+    return std::make_shared<AutoExecutableNetwork>(executableNetwork,
+                                                   selectedDevice,
+                                                   enablePerfCounters);
+}
+
+IE::QueryNetworkResult AutoInferencePlugin::QueryNetwork(const IE::CNNNetwork& network, const ConfigType& config) const {
+    IE::QueryNetworkResult queryResult = {};
+    if (GetCore() == nullptr) {
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    if (network.getFunction() == nullptr) {
+        IE_THROW() << "AUTO device supports just ngraph network representation";
+    }
+
+    auto fullConfig = mergeConfigs(_config, config);
+    auto metaDevices = GetDeviceChoice(fullConfig);
+    std::unordered_set<std::string> supportedLayers;
+    std::unordered_set<std::string> supportedDevices;
+    for (auto&& value : metaDevices) {
+        try {
+            auto deviceQr = GetCore()->QueryNetwork(network, value.deviceName, value.config);
+            std::unordered_set<std::string> deviceSupportedLayers;
+            for (auto &&layerQr : deviceQr.supportedLayersMap) {
+                deviceSupportedLayers.emplace(layerQr.first);
+            }
+            supportedLayers = supportedLayers.empty()
+                            ? deviceSupportedLayers : (deviceSupportedLayers.empty()
+                            ? supportedLayers : IE::details::Intersection(
+                                 supportedLayers, deviceSupportedLayers));
+            supportedDevices.insert(value.deviceName);
+            break;
+        } catch (...) {
+        }
+    }
+
+    if (supportedDevices.empty()) {
+        IE_THROW() << "Please, check environment due to no supported devices can be used";
+    }
+
+    for (auto&& supportedLayer : supportedLayers) {
+        queryResult.supportedLayersMap[supportedLayer] = GetName();
+    }
+    return queryResult;
+}
+
+IE::Parameter AutoInferencePlugin::GetConfig(const std::string& name,
+                                             const std::map<std::string, IE::Parameter> & options) const {
+    auto it = _config.find(name);
+    if (it == _config.end()) {
+        IE_THROW() << "Unsupported config key: " << name;
+    } else {
+        return { it->second };
+    }
+}
+
+void AutoInferencePlugin::SetConfig(const ConfigType& config) {
+    for (auto && kvp : config) {
+        _config[kvp.first] = kvp.second;
+    }
+}
```

Comment:
do we need to support SetConfig / GetConfig? Looks like we don't have any config keys for auto..

Reply:
yes we need it, because there are tests to call this function, the default will throw exception.